rosbag2_transport/src/rosbag2_transport/player.cpp

```
@@ -951,6 +988,27 @@ std::unordered_map<std::string,
   return topic_to_client_map;
 }
 
+std::unordered_map<std::string, std::shared_ptr<rclcpp_action::GenericClient>>
+PlayerImpl::get_action_clients()
+{
+  std::unordered_map<std::string, std::shared_ptr<rclcpp_action::GenericClient>>
+  topic_to_action_client_map;
+  for (const auto & [action_name, action_client] : action_clients_) {
+    topic_to_action_client_map[action_name] = action_client->generic_client();
+  }
+  return topic_to_action_client_map;
+}
+
+bool
+PlayerImpl::goal_handle_in_process(std::string action_name, const rclcpp_action::GoalUUID & goal_id)
+{
+  if (action_name.empty() || action_clients_.find(action_name) == action_clients_.end()) {
+    throw(std::invalid_argument("Action client for action '" + action_name + "' does not exist."));
+  }
```

i think `action_name.empty()` can be invalid argument and we can raise the exception.
but it is expected that goal_id might not be found on the list as API design? in that case, i would print the warning that it cannot find the goal_id on the list, and return false to avoid raising the exception for the application.

---

Okay. Update this.